Backpack/Dialog/Classes/BPKDialogControllerAnimator.m

```diff
@@ -0,0 +1,92 @@
+ * limitations under the License.
+ */
+#import "BPKDialogControllerAnimator.h"
+
+#import "BPKDialogController.h"
+
+NS_ASSUME_NONNULL_BEGIN
+@interface BPKDialogControllerAnimator()
+@property(nonatomic, assign, class, readonly) NSTimeInterval duration;
+@end
+
+@implementation BPKDialogControllerAnimator
+
+- (instancetype)init {
+    self = [super init];
+
+    if (self) {
+        self.presenting = YES;
+    }
+
+    return self;
+}
+
+- (void)animateTransition:(nonnull id<UIViewControllerContextTransitioning>)transitionContext {
+    UIView *containerView = transitionContext.containerView;
+    UIView *toView = [transitionContext viewForKey:UITransitionContextToViewKey];
+    UIViewController *_Nullable controller = self.isPresenting ? [transitionContext viewControllerForKey:UITransitionContextToViewControllerKey] : [transitionContext viewControllerForKey:UITransitionContextFromViewControllerKey];
+    BPKDialogController *dialogController = [controller isKindOfClass:[BPKDialogController class]] ? (BPKDialogController *)controller : nil;
+    NSAssert(dialogController, @"`UITransitionContextToViewControllerKey` for `BPKDialogControllerAnimator` should always be an instance of `BPKDialogController`");
+
+    if (self.isPresenting) {
+        [containerView addSubview:toView];
+        if (dialogController) {
+            [dialogController setScrimAlpha:0.0];
+        }
+        [controller.view layoutIfNeeded];
+
+        [UIView animateKeyframesWithDuration:[[self class] duration] delay:0.0 options:0 animations:^{
+            if (dialogController) {
+                [dialogController setScrimAlpha:1.0];
+                [dialogController addPresentingKeyFrameContentAnimationWithRelativeStartTime:0.4 relativeDuration:0.6];
+            }
+        } completion:^(BOOL finished) {
+            [transitionContext completeTransition:YES];
+        }];
+
+    } else {
+        if (dialogController) {
+            [dialogController setScrimAlpha:1.0];
+            [controller.view layoutIfNeeded];
+        }
+
+        [UIView animateKeyframesWithDuration:[[self class] duration] delay:0.0 options:0 animations:^{
+            if (dialogController && dialogController.style == BPKDialogControllerStyleAlert) {
+                dialogController.view.alpha = 0.0;
+            } else if (dialogController && dialogController.style == BPKDialogControllerStyleBottomSheet) {
+                [dialogController setScrimAlpha:0.0];
+                [dialogController addDismissingKeyFrameContentAnimationWithRelativeStartTime:0.6 relativeDuration:0.4];
+            }
+        } completion:^(BOOL finished) {
+            [transitionContext completeTransition:YES];
+        }];
+    }
+}
+
+- (NSTimeInterval)transitionDuration:(nullable id<UIViewControllerContextTransitioning>)transitionContext {
+    return [[self class] duration];
+}
+
+#pragma mark - Private
+
++ (NSTimeInterval)duration {
+    return 0.35;
```

Comment:
don't we have a backpack constant for this?

Reply:
Nope, this is based the default timings UIKit uses for their transitions.

Reply:
I reckon we should document that - quick comment above it or something? Principle of least surprise etc.